### PR TITLE
Fix soft button object invalid states

### LINKED ISSUE #1774: SoftButtonObject doesn't handle error states correctly
### Bug Report
The SoftButtonObject can be created with states that don't exist or are incorrect. We should check these states on construction to ensure that developers can't send incorrect soft buttons.

1. Creating a `SoftButtonObject` with empty states should throw an exception
2. Creating a `SoftButtonObject` with states where the initial state name is not among the set states.

##### Reproduction Steps
1. Create a soft button object with empty state list
2. Send the soft button object

##### Expected Behavior
The soft buttons should throw an exception

##### Observed Behavior
The soft buttons return a success because no soft buttons are sent in the `Show`

##### OS & Version Information
* Android Version: n/a
* SDL Android Version: v5.3.1
* Testing Against: Sync 3

##### Test Case, Sample Code, and / or Example App
n/a

### PR DESCRIPTION
**Pull request by noah-livio**
Fixes #1774

This PR is ready for review.

### Risk
This PR adds a new public getter function to SoftButtonObject

### Testing Plan
- [x] I have verified that I have not introduced new warnings in this PR
- [x] I have run the unit tests with this PR
- [x] I have tested this PR against Core and verified behavior
- [x] I have tested Android and Java SE

#### Unit Tests
 - `testConstructSoftButtonObjectWithEmptyStateList()` passes if `getAttemptedToAssignEmptyStateList()` returns true after an empty state list is passed to the main primary constructor and fails otherwise.

- `testAssignEmptyStateListToSoftButtonObject()` passes if an `getAttemptedToAssignEmptyStateList()` returns true after an empty state list is passed to `setStates()` of an instance of `SoftButtonObject` that already has at least one state in its state list. 

##### Tests performed by modifying a local clone of hello_sdl_android

- Constructed an instance of `SoftButtonObject` with a state list containing a single state, added it to a list of `SoftButtonObject`s, and attempted to display them to the Generic HMI using `ScreenManager`. Results: 
  - That instance along with all the other instances in the same transaction displayed successfully.
  - The app continued to function normally
- Constructed an instance of `SoftButtonObject` with an empty state list, added it to a list of `SoftButtonObject`s, and attempted to display them to the Generic HMI using `ScreenManager`. Results: 
   - The instance printed an error with `DebugTool` stating that the state list is empty
   - That instance and all other other instances within the same transaction failed to display to the HMI
   - The app otherwise continued to function normally
-  Constructed an instance of `SoftButtonObject` with a state list containing a single state, reassigned the state list to an empty list using `setStates()`, added it to a list of `SoftButtonObject`s, and attempted to display them to the Generic HMI using `ScreenManager`. Results:
   - The instance printed an error with `DebugTool` stating that the state list is empty
   - The instance did not set overwrite the old state list with the new state list
   - The app otherwise continued to function normally

SDL Core / 8.0.0 / master / 68f082169e0a40fccd9eb0db3c83911c28870f07 / module tested against: SDL Core on Ubuntu 20.04.3 LTS Virtual Machine
Generic HMI / 0.11.0 / master / 47e0ad42f05b27adff61befd864e79c2ab4b8cec / module tested against: Generic HMI on Ubuntu 20.04.3 LTS Virtual Machine

### Summary
This pull request adds an if statement to `SoftButtonObject.setStates()` that checks if the given state list is empty and displays  if it is. This PR also replaces the direct state list assignment in the main constructor with usage of `setStates()` to enforce the check in the constructor without writing redundant code.

##### Breaking Changes
* None

##### Enhancements
* Adds a getter function to check if an empty list was passed to the instance 

##### Bug Fixes
* Fixes #1774

### CLA
- [x] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

### CODE REVIEW

File: android/sdl_android/src/androidTest/java/com/smartdevicelink/managers/screen/SoftButtonManagerTests.java [hunk 1]
```diff
@@ -402,4 +404,32 @@ public void testSoftButtonStateEquals() {
...
+     */
+    @Test
+    public void testConstructSoftButtonObjectWithEmptyStateList() {
+        List<SoftButtonState> stateList = new ArrayList<>();
+        // Uncommenting the following lines should make the test fail
+        // SoftButtonState softButtonState = new SoftButtonState("object1-state1", "o1s1", null);
+        // stateList.add(softButtonState);
+        SoftButtonObject softButtonObject = new SoftButtonObject("hi", stateList, "Hi", null);
+
+        assertTrue(softButtonObject.getAttemptedToAssignEmptyStateList());
```
Comment (RHenigan) [inline android/sdl_android/src/androidTest/java/com/smartdevicelink/managers/screen/SoftButtonManagerTests.java:419]: Without the boolean flag we will need to change our assertions.
since we will return out of the constructor early most of our fields will still be null so we can use those for our assertions.
`assertNull(softButtonObject.getName());`
and if we make a SoftButtonObject with a non-empty state list we can assert the fields are set to what was passed in the constructor
`assertEquals("hi", softButtonObject.getName());`

File: android/sdl_android/src/androidTest/java/com/smartdevicelink/managers/screen/SoftButtonManagerTests.java [hunk 2]
```diff
@@ -402,4 +404,32 @@ public void testSoftButtonStateEquals() {
...
+     * Test assigning an empty state list to existing SoftButtonObject
+     */
+    @Test
+    public void testAssignEmptyStateListToSoftButtonObject() {
+        List<SoftButtonState> stateList = new ArrayList<>();
+        SoftButtonState softButtonState = new SoftButtonState("object1-state1", "o1s1", null);
+
+        SoftButtonObject softButtonObject = new SoftButtonObject("hi", softButtonState, null);
+        // Commenting the following line should make the test fail
+        softButtonObject.setStates(stateList);
```
Comment (RHenigan) [inline android/sdl_android/src/androidTest/java/com/smartdevicelink/managers/screen/SoftButtonManagerTests.java:432]: setStates should be returning early if it is passed an empty list, this means the states on the object will not be updated. to test this we can make an assertion before setting the state that the list is what is expected from the constructor.
`assertEquals(Collections.singletonList(softButtonState), softButtonObject.getStates());`
and then after trying to set the states to an empty list we can make the same assertion as setStates should return early and not update the field
`assertEquals(Collections.singletonList(softButtonState), softButtonObject.getStates());`

File: base/src/main/java/com/smartdevicelink/managers/screen/SoftButtonObject.java [hunk 3]
```diff
@@ -395,4 +404,13 @@ public SoftButtonObject clone() {
         }
         return null;
     }
+
+    /**
+     * Used to make unit testing easier by removing the need to read debug logs programmatically
+     *
+     * @return True if the last list passed to setStates() was empty and false otherwise
+     */
+    public boolean getAttemptedToAssignEmptyStateList() {
```
Comment (RHenigan) [inline base/src/main/java/com/smartdevicelink/managers/screen/SoftButtonObject.java:413]: This will not be needed if we remove our boolean flag.

**Review by RHenigan — CHANGES_REQUESTED** [review]

File: base/src/main/java/com/smartdevicelink/managers/screen/SoftButtonObject.java [hunk 4]
```diff
@@ -264,6 +266,13 @@ public void setName(@NonNull String name) {
      * @param states a list of the object's soft button states. <strong>states should be unique for every SoftButtonObject. A SoftButtonState instance cannot be reused for multiple SoftButtonObjects.</strong>
      */
     public void setStates(@NonNull List<SoftButtonState> states) {
+        // If the list of states is empty, throw an error with DebugTool and return
+        attemptedToAssignEmptyStateList = states.isEmpty();
+        if (attemptedToAssignEmptyStateList) {
+            DebugTool.logError(TAG,"The state list is empty");
+            return;
+        }
```
Comment (RHenigan) [inline base/src/main/java/com/smartdevicelink/managers/screen/SoftButtonObject.java:274]: Since we are doing it the constructor i think it also makes sense to check the list does not have duplicate state names here as well.
```
if (hasTwoStatesOfSameName(states)) {
    //Log error
    //return
}
```

Comment (noah-livio) [reply base/src/main/java/com/smartdevicelink/managers/screen/SoftButtonObject.java:274]: That sounds reasonable to me, I just wasn't sure I should increase the scope of this PR

File: android/sdl_android/src/androidTest/java/com/smartdevicelink/managers/screen/SoftButtonManagerTests.java (hunk 1 above)
Comment (noah-livio) [reply android/sdl_android/src/androidTest/java/com/smartdevicelink/managers/screen/SoftButtonManagerTests.java:419]: In the constructor based test, I think going with something like `assertNull(softButtonObject.getStates());` and `assertEquals(stateList, softButtonObject2.getStates());` for the error case and normal case, respectively, will be more resilient to future changes since it directly tests the actual state list. I hope that is an okay alternative to your suggestion.

Comment (RHenigan) [reply android/sdl_android/src/androidTest/java/com/smartdevicelink/managers/screen/SoftButtonManagerTests.java:419]: Wouldn't this test case fail when sending an empty list though?
In the constructor if the list is empty we should try to return early and none of the fields on that object will be set so those fields would be null?
If this is the case the assertion would fail because you are comparing an empty list to null.
`assertEquals(stateList, softButtonObject2.getStates())` stateList being an empty list and getStates() returning null.

You can write a second test case to test the event that a non-empty list is sent in the constructor. in this case yes we would want to verify getStates is returning the same list that was sent in the constructor

Comment (RHenigan) [reply android/sdl_android/src/androidTest/java/com/smartdevicelink/managers/screen/SoftButtonManagerTests.java:419]: Sorry, I misread your previous comment. Yes I think that is ok
If the list is empty we will return early and should verify that the states remain null.
If the list is populated then we should verify the object if created with the correct states list.

Comment (noah-livio) [reply android/sdl_android/src/androidTest/java/com/smartdevicelink/managers/screen/SoftButtonManagerTests.java:419]: It's all good, I'm just happy I'm going in the right direction in regard to them, I probably could have written my first comment more clearly to prevent misreading

**Comment by noah-livio** [comment]
Fixes #1774

This PR is ready for review.

### Risk
None

### Testing Plan
- [x] I have verified that I have not introduced new warnings in this PR
- [x] I have run the unit tests with this PR
- [x] I have tested this PR against Core and verified behavior
- [x] I have tested Android and Java SE

#### Unit Tests
- `testConstructSoftButtonObjectWithEmptyStateList()` passes if `SoftButtonObject.getStates()` returns null when the constructor is given an empty state list
- `testConstructingSoftButtonObjectWithNonEmptyStateList()` passes if `SoftButtonObject.getStates()` returns the nonempty state list passed to the constructor
- `testAssignNonEmptyStateListToSoftButtonObject()` passes if `SoftButtonObject.getStates()` returns a state list not equal to the empty state list passed to `SoftButtonObject.setStates()`
- `testAssignEmptyStateListToSoftButtonObject()` passes if `SoftButtonObject.getStates()` returns a state list equal to the nonempty state list passed to `SoftButtonObject.setStates()`


#### Core Tests

##### Tests performed by modifying a local clone of hello_sdl_android

- Constructed an instance of `SoftButtonObject` with a state list containing a single state, added it to a list of `SoftButtonObject`s, and attempted to display them to the Generic HMI using `ScreenManager`. Results: 
  - That instance along with all the other instances in the same transaction displayed successfully.
  - The app continued to function normally
- Constructed an instance of `SoftButtonObject` with an empty state list, added it to a list of `SoftButtonObject`s, and attempted to display them to the Generic HMI using `ScreenManager`. Results: 
   - The instance printed an error with `DebugTool` stating that the state list is empty
   - That instance and all other other instances within the same transaction failed to display to the HMI
   - The app otherwise continued to function normally
-  Constructed an instance of `SoftButtonObject` with a state list containing a single state, reassigned the state list to an empty list using `setStates()`, added it to a list of `SoftButtonObject`s, and attempted to display them to the Generic HMI using `ScreenManager`. Results:
   - The instance printed an error with `DebugTool` stating that the state list is empty
   - The instance did not overwrite the old state list with the new state list
   - The app otherwise continued to function normally

SDL Core / 8.0.0 / master / 68f082169e0a40fccd9eb0db3c83911c28870f07 / module tested against: SDL Core on Ubuntu 20.04.3 LTS Virtual Machine
Generic HMI / 0.11.0 / master / 47e0ad42f05b27adff61befd864e79c2ab4b8cec / module tested against: Generic HMI on Ubuntu 20.04.3 LTS Virtual Machine

### Summary
This pull request adds an if statement to `SoftButtonObject.setStates()` along with an identical if statement in the main constructor that throws an error and returns if the state list passed is empty

### Changelog
##### Breaking Changes
* None

##### Enhancements
* None

##### Bug Fixes
* Fixes #1774

### CLA
- [x] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

File: android/sdl_android/src/androidTest/java/com/smartdevicelink/managers/screen/SoftButtonManagerTests.java [hunk 5]
```diff
@@ -33,6 +33,7 @@
 import com.smartdevicelink.proxy.rpc.listeners.OnRPCNotificationListener;
 import com.smartdevicelink.test.Validator;
 
+import org.junit.Assert;
```
Comment (RHenigan) [inline android/sdl_android/src/androidTest/java/com/smartdevicelink/managers/screen/SoftButtonManagerTests.java:36]: We can remove this unused import

File: android/sdl_android/src/androidTest/java/com/smartdevicelink/managers/screen/SoftButtonManagerTests.java [hunk 6]
```diff
@@ -56,11 +58,14 @@
 import static org.mockito.Mockito.mock;
 import static org.mockito.Mockito.when;
 
+import android.database.sqlite.SQLiteCantOpenDatabaseException;
```
Comment (RHenigan) [inline android/sdl_android/src/androidTest/java/com/smartdevicelink/managers/screen/SoftButtonManagerTests.java:61]: we can remove this. unused import

File: android/sdl_android/src/androidTest/java/com/smartdevicelink/managers/screen/SoftButtonManagerTests.java [hunk 7]
```diff
@@ -56,11 +58,14 @@
...
+
 /**
  * This is a unit test class for the SmartDeviceLink library manager class :
  * {@link SoftButtonManager}
  */
 @RunWith(AndroidJUnit4.class)
+//@RunWith(JUnit4.class)
```
Comment (RHenigan) [inline android/sdl_android/src/androidTest/java/com/smartdevicelink/managers/screen/SoftButtonManagerTests.java:68]: We can remove this comment as well

File: android/sdl_android/src/androidTest/java/com/smartdevicelink/managers/screen/SoftButtonManagerTests.java [hunk 8]
```diff
@@ -402,4 +407,55 @@ public void testSoftButtonStateEquals() {
...
+     */
+    @Test
+    public void testAssignEmptyStateListToSoftButtonObject() {
+        List<SoftButtonState> stateList = new ArrayList<>();
+        SoftButtonState softButtonState = new SoftButtonState("object1-state1", "o1s1", null);
+
+        SoftButtonObject softButtonObject = new SoftButtonObject("hi", softButtonState, null);
+
+        softButtonObject.setStates(stateList);
+        assertNotEquals(stateList, softButtonObject.getStates());
```
Comment (RHenigan) [inline android/sdl_android/src/androidTest/java/com/smartdevicelink/managers/screen/SoftButtonManagerTests.java:444]: Rather than assert that these are not equal we could assert that the state on the Object is left unchanged after calling setStates.
`assertEquals(softButtonState, softButtonObject.getStates)`
You may want to use the constructor that takes a list for states rather than a single state. or you could make a list out of `softButtonState` before you make the assertion

File: android/sdl_android/src/androidTest/java/com/smartdevicelink/managers/screen/SoftButtonManagerTests.java [hunk 9]
```diff
@@ -402,4 +407,55 @@ public void testSoftButtonStateEquals() {
...
+    @Test
+    public void testAssignNonEmptyStateListToSoftButtonObject() {
+        List<SoftButtonState> stateList = new ArrayList<>();
+        SoftButtonState softButtonState = new SoftButtonState("object1-state1", "o1s1", null);
+
+        SoftButtonObject softButtonObject = new SoftButtonObject("hi", softButtonState, null);
+
+        stateList.add(softButtonState);
+        softButtonObject.setStates(stateList);
+        assertEquals(stateList, softButtonObject.getStates());
```
Comment (RHenigan) [inline android/sdl_android/src/androidTest/java/com/smartdevicelink/managers/screen/SoftButtonManagerTests.java:459]: For this test it may be beneficial to have stateList contain a state that is different than what was passed in the constructor. this way we are verifying the states are in fact being updated.

File: base/src/main/java/com/smartdevicelink/managers/screen/SoftButtonObject.java [hunk 10]
```diff
@@ -73,17 +73,22 @@
...
             DebugTool.logError(TAG, "Two states have the same name in states list for soft button object");
             return;
         }
 
         this.name = name;
-        this.states = states;
         this.currentStateName = initialStateName;
         this.buttonId = SOFT_BUTTON_ID_NOT_SET_VALUE;
         this.onEventListener = onEventListener;
+        this.states = states;
```
Comment (RHenigan) [inline base/src/main/java/com/smartdevicelink/managers/screen/SoftButtonObject.java:91]: Can we put this back to the same line it was previously on? this will help to reduce noise in the diff for changes that aren't really changes

File: base/src/main/java/com/smartdevicelink/managers/screen/SoftButtonObject.java [hunk 11]
```diff
@@ -395,4 +406,5 @@ public SoftButtonObject clone() {
         }
         return null;
     }
+
```
Comment (RHenigan) [inline base/src/main/java/com/smartdevicelink/managers/screen/SoftButtonObject.java:409]: just cleaning up whitespace

File: android/sdl_android/src/androidTest/java/com/smartdevicelink/managers/screen/SoftButtonManagerTests.java [hunk 12]
```diff
@@ -44,6 +45,7 @@
 import java.util.Collections;
 import java.util.List;
 
+
```
Comment (RHenigan) [inline android/sdl_android/src/androidTest/java/com/smartdevicelink/managers/screen/SoftButtonManagerTests.java:48]: just cleaning up white space.

File: base/src/main/java/com/smartdevicelink/managers/screen/SoftButtonObject.java (hunk 4 above)
Comment (RHenigan) [reply base/src/main/java/com/smartdevicelink/managers/screen/SoftButtonObject.java:276]: After looking at the iOS implementation the iOS team will also need to make changes. With the teams input I think it makes sense to add this check to setStates

File: base/src/main/java/com/smartdevicelink/managers/screen/SoftButtonObject.java (hunk 10 above)
Comment (noah-livio) [reply base/src/main/java/com/smartdevicelink/managers/screen/SoftButtonObject.java:91]: If you mean where it was when I was using setStates() in the constructor, I don't think I can without making the diff noisier, because now the check is happening there and the check obviously has to happen first to prevent assignment in the error case. This change is just reverting the states assignment back to how it was in develop before I forked. Maybe I could squash commits? But I think that might be more trouble than it's worth

Comment (joeygrover) [reply base/src/main/java/com/smartdevicelink/managers/screen/SoftButtonObject.java:91]: Just move `this.states = states;` back under `this.name= name;`

Comment (noah-livio) [reply base/src/main/java/com/smartdevicelink/managers/screen/SoftButtonObject.java:91]: Oh okay, will do

**Comment by noah-livio** [comment]
Fixes #1774

This PR is ready for review.

### Risk
None

### Testing Plan
- [x] I have verified that I have not introduced new warnings in this PR
- [x] I have run the unit tests with this PR
- [x] I have tested this PR against Core and verified behavior
- [x] I have tested Android and Java SE

#### Unit Tests
- `testConstructSoftButtonObjectWithEmptyStateList()` passes if `SoftButtonObject.getStates()` returns null when the constructor is given an empty state list
- `testConstructingSoftButtonObjectWithNonEmptyStateList()` passes if `SoftButtonObject.getStates()` returns the nonempty state list given to the constructor
- `testAssignNonEmptyStateListToSoftButtonObject()` passes if `SoftButtonObject.getStates()` returns the nonempty state list given to the constructor after `SoftButtonObject.setStates()` is given an empty state list
- `testAssignEmptyStateListToSoftButtonObject()` passes if `SoftButtonObject.getStates()` returns a state list equal to the nonempty state list passed to `SoftButtonObject.setStates()`


#### Core Tests

##### Tests performed by modifying a local clone of hello_sdl_android

- Constructed an instance of `SoftButtonObject` with a state list containing a single state, added it to a list of `SoftButtonObject`s, and attempted to display them to the Generic HMI using `ScreenManager`. Results: 
  - That instance along with all the other instances in the same transaction displayed successfully.
  - The app continued to function normally
- Constructed an instance of `SoftButtonObject` with an empty state list, added it to a list of `SoftButtonObject`s, and attempted to display them to the Generic HMI using `ScreenManager`. Results: 
   - The instance printed an error with `DebugTool` saying that the state list is empty
   - That instance and all other other instances within the same transaction failed to display to the HMI
   - The app otherwise continued to function normally
-  Constructed an instance of `SoftButtonObject` with a state list containing a single state, reassigned the state list to an empty list using `setStates()`, added it to a list of `SoftButtonObject`s, and attempted to display them to the Generic HMI using `ScreenManager`. Results:
   - The instance printed an error with `DebugTool` saying that the state list is empty
   - The instance did not overwrite the old state list with the new state list
   - The app otherwise continued to function normally
-  Constructed an instance of `SoftButtonObject` with a state list containing a single state, reassigned the state list to a state list with states with duplicate names using `setStates()`, added it to a list of `SoftButtonObject`s, and attempted to display them to the Generic HMI using `ScreenManager`. Results:
   - The instance printed an error with `DebugTool` saying "Two states have the same name in states list for soft button object"
   - The instance did not overwrite the old state list with the new state list
   - The app otherwise continued to function normally

SDL Core / 8.0.0 / master / 68f082169e0a40fccd9eb0db3c83911c28870f07 / module tested against: SDL Core on Ubuntu 20.04.3 LTS Virtual Machine
Generic HMI / 0.11.0 / master / 47e0ad42f05b27adff61befd864e79c2ab4b8cec / module tested against: Generic HMI on Ubuntu 20.04.3 LTS Virtual Machine

### Summary
This pull request adds an if statement to `SoftButtonObject.setStates()` along with an identical if statement in the main constructor that throws an error and returns if the state list passed is empty. This pull request also adds a similar check to `SoftButtonObject.setStates()` for state lists containing states with duplicate names.

### Changelog
##### Breaking Changes
* None

##### Enhancements
* None

##### Bug Fixes
* Fixes #1774

### CLA
- [x] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

File: base/src/main/java/com/smartdevicelink/managers/screen/SoftButtonObject.java [hunk 13]
```diff
@@ -73,6 +71,11 @@
      */
     public SoftButtonObject(@NonNull String name, @NonNull List<SoftButtonState> states, @NonNull String initialStateName, OnEventListener onEventListener) {
 
+        // If the list of states is empty, throw an error with DebugTool and return
+        if (states.isEmpty()) {
+            DebugTool.logError(TAG,"The state list is empty");
+            return;
+        }
```
Comment (joeljfischer) [inline base/src/main/java/com/smartdevicelink/managers/screen/SoftButtonObject.java:78]: Alignment note:

On iOS I did this

```objc
BOOL hasStateWithInitialName = NO;
for (SDLSoftButtonState *state in states) {
    if ([state.name isEqualToString:initialStateName]) {
        hasStateWithInitialName = YES;
    }
}
NSAssert(![self sdl_hasTwoStatesOfSameName:states], @"A SoftButtonObject must have states with different names.");
NSAssert(hasStateWithInitialName, @"A SoftButtonObject must have a state with initialStateName.");
if ([self sdl_hasTwoStatesOfSameName:states]) { return nil; }
if (!hasStateWithInitialName) { return nil; }
```

The advantage this has is only throwing an exception in DEBUG builds (I don't know that is available on Java Suite). Second, this tests not only that there _are_ states, but also that the initial state name is available. The for loop won't run if there are no states, so the boolean will just stay false. That check at least should be a part of this PR as well.

Last, I also made a change to `SoftButtonState` on iOS to check if both text and image are `null`, which is an invalid state configuration. I think that should be a part of this PR as well.

File: base/src/main/java/com/smartdevicelink/managers/screen/SoftButtonObject.java (hunk 4 above)
Comment (joeljfischer) [reply base/src/main/java/com/smartdevicelink/managers/screen/SoftButtonObject.java:274]: See my above comment

**Review by joeljfischer — CHANGES_REQUESTED** [review]

File: base/src/main/java/com/smartdevicelink/managers/screen/SoftButtonObject.java (hunk 13 above)
Comment (noah-livio) [reply base/src/main/java/com/smartdevicelink/managers/screen/SoftButtonObject.java:78]: Understood, I'll work on implementing this

**Comment by noah-livio** [comment]
I have added the `initialStateName` check in the constructor, added `AssertionError`s that only throw when built in debug mode, and adjusted existing tests to avoid failing for the `initialStateName` check. I also added a debug only `AssertionError` to `SoftButtonState` in the check for null text and image.

**Comment by RHenigan** [comment]
Older unit tests fail as a result of the new hasStateWithInitialName check.
Some tests are using empty lists and null or mismatched initia state names which are the states we are trying to avoid.

ScreenManagerTests.testAssigningIdsToSoftButtonObjects
SoftButtonManagerTests.testSoftButtonObjectEquals

**Review by RHenigan — CHANGES_REQUESTED** [review]

**Review by RHenigan — CHANGES_REQUESTED** [review]
`testAssignNonEmptyStateListToSoftButtonObject` is now failing, this test would need to set a list with a state that is equivalent to the currentState. We should also test to verify the negative is true as well (confirm setting a list without the currentState fails to update)

File: base/src/main/java/com/smartdevicelink/managers/screen/SoftButtonObject.java [hunk 14]
```diff
@@ -264,6 +275,26 @@ public void setName(@NonNull String name) {
...
+
+        if (repeatedStateNames) {
+            DebugTool.logError(TAG, "A SoftButtonObject must have states with different names.");
+            return;
+        }
+
+        boolean hasStateWithCurrentName = false;
+        for (SoftButtonState state : states) {
+            if(state.getName().equals(currentStateName)) {
+                hasStateWithCurrentNameName = true;
```
Comment (RHenigan) [inline base/src/main/java/com/smartdevicelink/managers/screen/SoftButtonObject.java:289]:
```suggestion
                hasStateWithCurrentName = true;
```

Comment (noah-livio) [reply base/src/main/java/com/smartdevicelink/managers/screen/SoftButtonObject.java:289]: Good catch, and that'll be easy to fix

Comment (RHenigan) [reply base/src/main/java/com/smartdevicelink/managers/screen/SoftButtonObject.java:289]: Also to avoid confusion for the developer we will still want to set the states in the case the currentState is not present in the list, but we will still want to log the error.

We should test and confirm that the SoftButtonManager can handle this error case (trying to upload a SoftButton where the currentState is not in the list)

**Comment by noah-livio** [comment]
Fixes #1774

This PR is ready for review.

### Risk
None

### Testing Plan
- [x] I have verified that I have not introduced new warnings in this PR
- [x] I have run the unit tests with this PR
- [x] I have tested this PR against Core and verified behavior
- [x] I have tested Android and Java SE

#### Unit Tests
- `testConstructSoftButtonObjectWithEmptyStateList()` passes if `SoftButtonObject.getStates()` returns null when the constructor is given an empty state list
- `testConstructingSoftButtonObjectWithNonEmptyStateList()` passes if `SoftButtonObject.getStates()` returns the nonempty state list given to the constructor
- `testConstructSoftButtonObjectWithInvalidInitialStateName()` passes if `SoftButtonObject.getStates()` returns null when the constructor is given a state list that does not contain `initialStateName`
- `testAssignStateListWithCurrentStateToSoftButtonObject()` passes if `SoftButtonObject.getStates()` returns the new state list given to `SoftButtonObject.setStates()`
- `testAssignStateListWithoutCurrentStateToSoftButtonObject()` passes if `SoftButtonObject.getStates()` returns the new state list given to `SoftButtonObject.setStates()`
- `testAssignEmptyStateListToSoftButtonObject()` passes if `SoftButtonObject.getStates()` returns a state list equal to the nonempty state list passed to `SoftButtonObject.setStates()`

#### Core Tests

##### Tests performed by modifying a local clone of hello_sdl_android

- Constructed an instance of `SoftButtonObject` with a state list containing a single state, added it to a list of `SoftButtonObject`s, and attempted to display them to the Generic HMI using `ScreenManager`. Results: 
  - That instance along with all the other instances in the same transaction displayed successfully.
  - The app continued to function normally
- Constructed an instance of `SoftButtonObject` with an empty state list, added it to a list of `SoftButtonObject`s, and attempted to display them to the Generic HMI using `ScreenManager`. Results: 
   - The instance printed an error with `DebugTool` saying that the state list is empty
   - That instance and all other other instances within the same transaction failed to display to the HMI
   - The app otherwise continued to function normally
-  Constructed an instance of `SoftButtonObject` with a state list containing a single state, reassigned the state list to an empty list using `setStates()`, added it to a list of `SoftButtonObject`s, and attempted to display them to the Generic HMI using `ScreenManager`. Results:
   - The instance printed an error with `DebugTool` saying that the state list is empty
   - The instance did not overwrite the old state list with the new state list
   - The app otherwise continued to function normally
-  Constructed an instance of `SoftButtonObject` with a state list containing a single state, reassigned the state list to a state list with states with duplicate names using `setStates()`, added it to a list of `SoftButtonObject`s, and attempted to display them to the Generic HMI using `ScreenManager`. Results:
   - The instance printed an error with `DebugTool` saying "Two states have the same name in states list for soft button object"
   - The instance did not overwrite the old state list with the new state list
   - The app otherwise continued to function normally

SDL Core / 8.0.0 / master / 68f082169e0a40fccd9eb0db3c83911c28870f07 / module tested against: SDL Core on Ubuntu 20.04.3 LTS Virtual Machine
Generic HMI / 0.11.0 / master / 47e0ad42f05b27adff61befd864e79c2ab4b8cec / module tested against: Generic HMI on Ubuntu 20.04.3 LTS Virtual Machine

### Summary
This pull request adds an if statement to `SoftButtonObject.setStates()` along with an identical if statement in the main constructor that throws an error and returns if the state list passed is empty. This pull request also adds a similar check to `SoftButtonObject.setStates()` for state lists containing states with duplicate names.

Additionally `AssertionError`s are now thrown in these error checks when built in debug mode.

### Changelog
##### Breaking Changes
* None

##### Enhancements
* None

##### Bug Fixes
* Fixes #1774

### CLA
- [x] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

**Review by RHenigan — APPROVED** [review]

**Comment by RHenigan** [comment]
Awaiting iOS Alignment before merging

File: base/src/main/java/com/smartdevicelink/managers/screen/SoftButtonObject.java [hunk 15]
```diff
@@ -264,6 +275,30 @@ public void setName(@NonNull String name) {
...
+
+        boolean hasStateWithCurrentName = false;
+        for (SoftButtonState state : states) {
+            if(state.getName().equals(currentStateName)) {
+                hasStateWithCurrentName = true;
+                break;
+            }
+        }
+        if (!hasStateWithCurrentName) {
+            DebugTool.logError(TAG, "A SoftButtonObject should have a state with currentStateName.");
```
Comment (joeljfischer) [inline base/src/main/java/com/smartdevicelink/managers/screen/SoftButtonObject.java:299]:
```suggestion
            DebugTool.logError(TAG, "A SoftButtonObject setting states must contain a state with currentStateName.");
```

Can you also include the name of `currentStateName`?

**Review by joeljfischer — CHANGES_REQUESTED** [review]
This also seems to be missing the changes to `SoftButtonState` to ensure a state never has neither text nor an image.

**Comment by noah-livio** [comment]
> This also seems to be missing the changes to `SoftButtonState` to ensure a state never has neither text nor an image.

That check is already handled in the constructor of `SoftButtonState`

**Review by joeljfischer — APPROVED** [review]

**Review by RHenigan — APPROVED** [review]
